x/pairing/keeper/msg_server_relay_payment.go

```diff
@@ -392,3 +391,102 @@ func appendRelayPaymentDetailsToEvent(from map[string]string, uniqueIdentifier u
+			utils.Attribute{Key: "chainID", Value: chainID},
+		)
+	}
+
+	stakeEntry, found, stakeIndex := k.epochStorageKeeper.GetStakeEntryByAddressFromStorage(ctx, stakeStorage, providerAddr)
+	if !found {
+		return utils.LavaFormatError("could not calculate provider reward by delegations", fmt.Errorf("provider stake entry not found"),
+			utils.Attribute{Key: "provider", Value: providerAddr},
+			utils.Attribute{Key: "chainID", Value: chainID},
+			utils.Attribute{Key: "totalRewards", Value: totalReward},
+		)
+	}
+
+	providerReward, err := k.calculateProviderDelegatorsRewards(ctx, stakeEntry, stakeIndex, totalReward)
+	if err != nil {
+		return utils.LavaFormatError("cannot calculate provider reward and update delegator reward map", err)
+	}
+
+	providerRewardCoins := sdk.Coins{sdk.NewCoin(epochstoragetypes.TokenDenom, providerReward)}
+	err = k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, providerAddr, providerRewardCoins)
+	if err != nil {
+		// panic:ok: reward transfer should never fail
+		utils.LavaFormatPanic("critical: failed to send reward to provider", err,
+			utils.Attribute{Key: "provider", Value: providerAddr},
+			utils.Attribute{Key: "reward", Value: providerRewardCoins},
+		)
+	}
+
+	delegations, err := k.dualStakingKeeper.GetProviderDelegators(ctx, providerAddr.String(), epoch)
+	if err != nil {
+		return utils.LavaFormatError("cannot get provider's delegators", err)
+	}
+
+	var relevantDelegations []dualstakingtypes.Delegation
+	for _, d := range delegations {
+		if d.ChainID == chainID {
+			relevantDelegations = append(relevantDelegations, d)
+		}
+	}
+
+	err = k.updateDelegatorsReward(ctx, stakeEntry, relevantDelegations, totalReward)
+	if err != nil {
+		return utils.LavaFormatError("cannot update delegators reward map", err)
+	}
+
+	return nil
+}
+
+// calculateProviderDelegatorsRewards returns the provider reward amount
+func (k Keeper) calculateProviderDelegatorsRewards(ctx sdk.Context, stakeEntry epochstoragetypes.StakeEntry, stakeIndex uint64, totalReward math.Int) (math.Int, error) {
+	// sanity check
+	if stakeEntry.DelegateCommission > 100 {
+		invalidCommission := stakeEntry.DelegateCommission
+		stakeEntry.DelegateCommission = 100
+		k.epochStorageKeeper.ModifyStakeEntryCurrent(ctx, stakeEntry.Chain, stakeEntry, stakeIndex)
+		utils.LavaFormatError("changed provider delegation commission to 100%", fmt.Errorf("delegation commission above max value"),
+			utils.Attribute{Key: "provider", Value: stakeEntry.Address},
+			utils.Attribute{Key: "chainID", Value: stakeEntry.Chain},
+			utils.Attribute{Key: "invalidCommission", Value: invalidCommission},
+		)
+	}
+
+	return k.dualStakingKeeper.CalcProviderReward(stakeEntry, totalReward), nil
+}
+
+// updateDelegatorsReward updates the delegator rewards map
+func (k Keeper) updateDelegatorsReward(ctx sdk.Context, stakeEntry epochstoragetypes.StakeEntry, delegations []dualstakingtypes.Delegation, totalReward math.Int) error {
```

Comment:
I think here is unsuitable to do this work. 
It would make more sense for dualstaking to loop through the delegations and do this work, there. It already has all the know-how, just needs to know the totalReward. And then none of those internals of dualstaking needs to be exported to here. (For instance, how provider/chainID/delegator are encoded into keys is none of this pairing-module's business).

Reply:
done. see https://github.com/lavanet/lava/commit/520bc5955430cf6ec1125c970dc5e3ab3aca6560